### PR TITLE
[9076] Remove CE chunked header from t.w.http CONNECT response

### PR DESCRIPTION
**Pull request by izzy**
According to [RFC 7231 Section 4.3.6](https://tools.ietf.org/html/rfc7231#page-31) "A server MUST NOT send any Transfer-Encoding or Content-Length header fields in a 2xx (Successful) response to CONNECT"

Fix for [Trac ticket 9076](https://twistedmatrix.com/trac/ticket/9076#ticket)

### CODE REVIEW

**Comment by exarkun** [comment]
Thanks.  Please add test coverage and a news fragment.

**Comment by izzy** [comment]
Are there any resources on how to add those tests? I'm new to developing twisted itself and kinda lost here what to do and how to do it. Especially got a problem with the fact that twisted does not fully support CONNECT requests. Any input on how to get started with this?

**Comment by izzy** [comment]
It seems that codecov report is a "little" off? Well anyway, added a test, added the topfile, as far as I can tell that should be it.

**Comment by exarkun** [comment]
> It seems that codecov report is a "little" off? Well anyway, added a test, added the topfile, as far as I can tell that should be it.

Yea, the codecov reports aren't really useful because there is so much random variation.  Sorry about that.  Thanks for adding the test - sorry I didn't get a chance to give you any hints on that.

**Comment by glyph** [comment]
While codecov definitely has accuracy issues, in _this_ case it's just reporting a lower number because it reports combined coverage, and no project member has pushed this change to a branch so the buildbots can run the full suite yet.  (This is also why the OS X builder is showing an unfinished status).

File: src/twisted/web/http.py
Comment (njsmith) [inline src/twisted/web/http.py]: `Content-Length` is totally unobjectionable in, say, a 4xx response to a CONNECT. It's only 2xx responses that are special.

File: src/twisted/web/http.py
Comment (njsmith) [inline src/twisted/web/http.py]: Note that you need `//` here on python 3.

**Comment by wsanchez** [comment]
Test pass except for missing news fragment.  Instructions for those are here: https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

**Comment by izzy** [comment]
@wsanchez I added already one.. not sure why that's triggering as error...?

**Comment by rodrigc** [comment]
@mkzero you submitted your change earlier.  After you submitted your change, unfortunately the process for newsfiles changes, *after* you submitted your change.
What you need to do is **git mv** one of your files from:
**src/twisted/web/topfiles/9076.bugfix** to **src/twisted/web/newsfragments/9076.bugfix**.

After you do that, commit, and push the change.

**Comment by rodrigc** [comment]
Also, before  you push your change to this branch, can you merge latest trunk into your branch?
There have been a lot of changes.

**Comment by izzy** [comment]
So I fixed a few things merged trunk into my branch, but I'm not sure if the error remaining in [Travis](https://travis-ci.org/twisted/twisted/jobs/283736825#L3153) is my fault..?

**Comment by rodrigc** [comment]
It's not your fault.  Travis started using Python 3.6.3, and something changed there causing test_sslverify.py to break

**Comment by izzy** [comment]
@rodrigc so AFAICS my PR should be mergeable?

**Comment by rodrigc** [comment]
This cannot be merged, because the failing Travis build will block this.
This is affecting all PR's, not just yours.

@mkzero Can you code review this: by clicking the green *Review changes* button at https://github.com/twisted/twisted/pull/910/files 

I need someone to code review that so I can merge it into trunk.  If I can get this fix into trunk, we can merge trunk into your branch, and get your branch passing Travis again.

**Comment by rodrigc** [comment]
@mkzero I got someone else to code review https://github.com/twisted/twisted/pull/910 and it has been merged to trunk.  Can you merge Twisted trunk into your branch?  That will get Travis passing on your branch.

**Comment by izzy** [comment]
Merged trunk into my branch but it seems the buildbot check hangs?

**Comment by rodrigc** [comment]
That's normal, since only a Twisted committer can trigger all the buildbots.

File: src/twisted/web/test/test_http.py
Comment (rodrigc) [inline src/twisted/web/test/test_http.py]: Please add docstring here to eliminate error here: https://travis-ci.org/twisted/twisted/jobs/284299844.
When writing docstring, follow this guideline: https://jml.io/pages/test-docstrings.html
and don't put **Test** in the description.

**Comment by izzy** [comment]
Seems like the tests finally are all green. Anything else holding it back from being merged into trunk?

**Comment by Aurora-Beta** [comment]
I'm affected by the issue as well. Can anyone say when i can expect this to be merged?

**Comment by izzy** [comment]
Can someone trigger the buildbot and merge this if possible? I've been waiting almost a year to get this bugfix into Twisted now...

**Comment by exarkun** [comment]
Triggered buildbot

**Comment by exarkun** [comment]
The fedeora24-py3.5 failure is unrelated to this PR so it shouldn't block the merge.  It's the known issue - https://twistedmatrix.com/trac/ticket/9382

File: src/twisted/web/test/test_http.py
Comment (exarkun) [inline src/twisted/web/test/test_http.py]: This assertion looks redundant with the following.  Is there a particular reason to have it?  If not, I think it should be removed.  The `assertResponseEquals` will cover this check.

File: src/twisted/web/test/test_http.py
Comment (exarkun) [inline src/twisted/web/test/test_http.py]: You can leave off "correct" here.  Tests are always for correct behavior. :wink:

File: src/twisted/web/test/test_http.py
Comment (exarkun) [inline src/twisted/web/test/test_http.py]: `An HTTP ...`

File: src/twisted/web/test/test_http.py
Comment (exarkun) [inline src/twisted/web/test/test_http.py]: `... neither of which is allowed in a 200 response.` would be more natural English.

File: src/twisted/web/http.py
Comment (exarkun) [inline src/twisted/web/http.py]: There should be space surrounding the `//` operator.

File: src/twisted/web/http.py
Comment (exarkun) [inline src/twisted/web/http.py]: There should be space surrounding the `//` operator.

File: src/twisted/web/newsfragments/9076.bugfix
Comment (exarkun) [inline src/twisted/web/newsfragments/9076.bugfix]: Please add a full stop (`.`) to the end of this sentence.

File: src/twisted/web/http.py
Comment (exarkun) [inline src/twisted/web/http.py]: It looks like the logging style in this module has changed in the interim.  Can you make this `self._log.info(...)` instead (I would almost say something about not putting "Warning" there but I see that other code in this module emits info messages that start with "Warning" so ...)

File: src/twisted/web/http.py
Comment (exarkun) [inline src/twisted/web/http.py]: is it "*should* not" or "*must* not"?  seems like it should be one or the other.

**Review by exarkun — APPROVED** [review]
Thanks.  Sorry about the long review delay.  I've left some minor comments inline.  I'd like to see them addressed but then I think this can be merged.

**Comment by izzy** [comment]
@exarkun I've implemented the changes, tests seem to run fine, can you run the buildbot again?

**Comment by twm** [comment]
I merged this forward in #1358. Closing this one as the original fork has disappeared, so it can't go anywhere.

**Comment by izzy** [comment]
Just for completeness sake, I didn't delete the fork per se, I guess Github just didn't handle me changing my username properly, sorry if that caused any inconveniences.

**Comment by twm** [comment]
@izzy Thanks for clarifying!

**Pull request closed without merging**
